src/troute-network/troute/DataAssimilation.py

```diff
@@ -1514,4 +1536,60 @@ def assemble_rfc_dataframes(rfc_timeseries_df, rfc_lake_gage_crosswalk, t0, rfc_
     reservoir_rfc_param_df['update_time'] = 0
     reservoir_rfc_param_df['rfc_persist_days'] = rfc_parameters.get('reservoir_rfc_forecast_persist_days', 11)
 
-    return reservoir_rfc_df, reservoir_rfc_param_df
\ No newline at end of file
+    return reservoir_rfc_df, reservoir_rfc_param_df
+
+def _read_lastobs_file(
+        lastobsfile,
+        station_id = "stationId",
+        ref_t_attr_id = "modelTimeAtOutput",
+        time_idx_id = "timeInd",
+        obs_discharge_id = "discharge",
+        discharge_nan = -9999.0,
+        time_shift = 0,
+        ):
+    with xr.open_dataset(lastobsfile) as ds:
+        gages = ds[station_id].values
+        
+        ref_time = datetime.strptime(ds.attrs[ref_t_attr_id], "%Y-%m-%d_%H:%M:%S")
+        
+        last_ts = ds[time_idx_id].values[-1]
+        
+        df_discharge = (
+            ds[obs_discharge_id].to_dataframe().                 # discharge to MultiIndex DF
+            replace(to_replace = discharge_nan, value = np.nan). # replace null values with nan
+            unstack(level = 0)                                   # unstack to single Index (timeInd)    
+        )
+        
+        last_obs_index = (
+            df_discharge.
+            apply(pd.Series.last_valid_index).                   # index of last non-nan value, each gage
+            to_numpy()                                           # to numpy array
```

Comment:
For some reason, `last_obs_index` returns all zeros for me (changing all discharge values to zero), however, `df_discharge` is not zero.

To find the last valid value, maybe consider something like this:

```python
df_discharge = ds[obs_discharge_id].to_dataframe().replace(to_replace=discharge_nan, value=np.nan).reset_index(level='timeInd', drop=True).stack().dropna()
last_obs_value = df_discharge.iloc[-1]
```
And for the index of the last value:

```python
last_obs_index = df_discharge[df_discharge == last_obs_value].idxmax()[0]
```

Reply:
I guess last_obs_index is only used for row index of df_discharge, being nothing to do with make discharge zero.  If you look into the final product, which is lastobs_df, the dataframe has non-zero lastobs_discharge values.